### DISCUSSION TITLE
Use Fish as applications launcher

### DISCUSSION THREAD

**Discussion by sulivanShu** [fish-shell/fish-shell#9327] [Ideas]
Hello, I'm trying to convert fish to an application launcher. To do this, I create the file:

$XDG_CONFIG_HOME/fish/functions/app_launcher.fish
```
function fish_app_launcher --on-event fish_preexec --description 'Fish App Launcher. Run a command and exit.' ;
	echo $argv | xargs -Iargument nohup fish -c "nohup argument > /dev/null &; exit" > /dev/null;
	exit ;
end
```
The function will intercept the command before it is executed, run it with nohup, and exit the terminal immediately.

I don't know why nohup is needed for fish. Without nohup, sometimes fish exit before the command is run. I would be interested in an explanation.

Then I created the small script:
```
sh -c "cat $XDG_CONFIG_HOME/fish/functions/app_launcher.fish | xargs --null -Iargument st -f 'Liberation Mono:size=12' -e fish -C 'argument'"
```
Which I can launch via a keyboard key ("Super" for example).

"st" is my terminal emulator.

There is no bug? Maybe a suggestion?

**Comment by ridiculousfish** [comment]
I don't know why two nohups are needed here, but the level of nesting is very deep: we have `sh` launching `xargs` launching `st` launching `fish` launching `xargs` launching `nohup` launching `fish`. It's not surprising something is getting mixed signals here, so to speak.

I would recommend trying to simplify this. It looks like you want to read a command and then execute it in the background; you can do that using `read`:

    read cmd
    eval $cmd &

this avoids trying to intercept it in `fish_preexec` which isn't really designed for that sort of thing.

**Reply by ridiculousfish** [reply]
Alternatively even simpler:

    read -t cmd
    $cmd &

**Comment by sulivanShu** [comment]
I managed to simplify the function:
```
function fish_app_launcher --on-event fish_preexec
	nohup fish -c "nohup $argv & exit"
	exit
end
```
But I don't understand how to do without fish_preexec.

If I use `read`, I won't have any of the functionality of fish in the app launcher.

To make things more clear, here is the nesting of the processes in htop:

```
openbox --config-file $XDG_CONFIG_HOME/openbox/lxde-rc.xml
└─ bash│sh -c cat $XDG_CONFIG_HOME/fish/functions/app_launcher.fish | xargs -0 -Iargument st -f 'Liberation Mono:size=12' -e fish -C 'argument'
   └─ xargs -0 -Iargument st -f Liberation Mono:size=12 -e fish -C argument
      └─ st -f Liberation Mono:size=12 -e fish -C function fish_app_launcher --on-event fish_preexec ?nohup fish -c "nohup $argv & exit" ?exit  end
         └─ fish -C function fish_app_launcher --on-event fish_preexec ?nohup fish -c "nohup $argv & exit" ?exit  end
```

- openbox intercepts the "super" keypress which triggers the script
- sh which launches the terminal
- st which launches the shell
- fish with the config file
- app_launcher.fish

When I type a command in this terminal and press "enter", app_launcher.fish intercepts the command, runs it in the background & nohup, and exits the terminal before the "normal" execution of the command.
